src/category_theory/reflect_isomorphisms.lean

```diff
@@ -0,0 +1,95 @@
+/-
+Copyright (c) 2020 Bhavik Mehta. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Bhavik Mehta
+-/
+import category_theory.limits.limits
+import category_theory.limits.preserves
+
+open category_theory category_theory.limits
+
+namespace category_theory
+
+universes v₁ v₂ u₁ u₂
+
+variables {C : Type u₁} [𝒞 : category.{v₁} C]
+include 𝒞
+
+section reflects_iso
+variables {D : Type u₂} [𝒟 : category.{v₂} D]
+include 𝒟
+
+/--
+Define what it means for a functor `F : C ⥤ D` to reflect isomorphisms: for any
+morphism `f : A ⟶ B`, if `F.map f` is an isomorphism then `f` is as well.
+Note that we do not assume or require that `F` is faithful.
+-/
+class reflects_isomorphisms (F : C ⥤ D) :=
+(reflects : Π {A B : C} (f : A ⟶ B) [is_iso (F.map f)], is_iso f)
+
+-- TODO: reflects iso is equivalent to reflecting limits of shape 1 (punit)
+
+-- Having this as an instance seems to break resolution, so let's not.
```

Comment:
I guess every `is_iso _` search will immediately create a new metavariable for a functor `F`, and then it either searches for `reflects_isomorphisms F` (which would be okay, because it should fail), or searches for `is_iso (F.map f)` (which would be bad, because then we'd try this same instance again and again.

I _think_ that we recently changed the behaviour in Lean to "start from the end" when multiple instances were required, on the principle that the "last" instances would typically be the hardest to satisfy.

We could verify this behaviour by turning on `trace class.instances true` in an example here, and checking that this is the loop that's occurring.

Alternatively, you could just swap the order of the two typeclasses here, and hope that solves the problem...

Reply:
In any case it should be a low priority instance.

Reply:
I think I'll leave this for now, it seems like it'll create many more issues than it would solve.

Reply:
The comment should be removed or explained, then.